### PR TITLE
Post request.completed only after envelopes from that request were added to a filter

### PR DESCRIPTION
**Pull request by dshulyak**
We are receiving request.completed from a peer only when all envelopes were already sent. Then we send them on the same thread in same order. Once all envelopes were processed we will process event and send it to the feed. This way caller will be sure that all messages from this request are already in the filter, and once he received request.completed he can fetch messages and process them, and after that complete request.

### CODE REVIEW

**Comment by adambabik** [comment]
LGTM but unit tests are missing.

File: whisperv6/whisper.go [hunk 1]
```diff
@@ -90,9 +90,9 @@ type Whisper struct {
 	peerMu sync.RWMutex       // Mutex to sync the active peer set
 	peers  map[*Peer]struct{} // Set of currently active peers
 
-	messageQueue chan *Envelope // Message queue for normal whisper messages
-	p2pMsgQueue  chan *Envelope // Message queue for peer-to-peer messages (not to be forwarded any further)
-	quit         chan struct{}  // Channel used for graceful exit
+	messageQueue chan *Envelope   // Message queue for normal whisper messages
+	p2pMsgQueue  chan interface{} // Message queue for peer-to-peer messages (not to be forwarded any further)
```
Comment (mandrigin) [inline whisperv6/whisper.go]: I would update the comment that is is either an envelope or an event now.

**Comment by dshulyak** [comment]
This solution broken as well. There are as many threads fetching messages from p2pMessageQueue as cores. So there is no guarantee of any ordering.

**Comment by dshulyak** [comment]
P2P envelopes and events now processed in single thread. I simply can't find better solution.

**Comment by mandrigin** [comment]
@dshulyak you can also introduce a intermetiate queue to handle those sequentially.

So original design is `network -> queue -> N threads`, which can be replaces with `network -> queue1 -> p2p event monitor -> queue 2 -> N threads`

essentially what p2p event monitor does is 

```
p2pEvent <- queue1
handleEvent
queue2 <- p2pEvent
```

**Comment by dshulyak** [comment]
> @dshulyak you can also introduce a intermetiate queue to handle those sequentially.

maybe i am missing a point.. but the goal is to send a request.completed event to feed after all envelopes preceding this event were added to filters

**Comment by mandrigin** [comment]
@dshulyak ah, okay, I missed the point then. So they are added to the filters in the multi-threaded way? then probably it is okay to just make it single-threaded for the time being.

**Pull request merged**